katsdpimageutils/primarybeamcorrection/primary_beam_correction.py

```diff
@@ -0,0 +1,250 @@
+    output: A separation vector between phase centre and source positions in radians.
+    """
+    p2w = image_wcs.pixel_to_world(row, col, 0, 0)[0]
+    separation_rad = p2w.separation(phase_center).rad
+    return separation_rad
+
+
+def central_freq(path):
+    """
+    Function to get central frequency of each frequency plane.
+    Input : image header
+    Output : List of central frequency in MHz of each frequency plane.
+    """
+    images = read_fits(path)
+    c_freq_plane = []
+    for i in range(1, images.header['NSPEC']+1):
+        c_freq_plane.append(images.header['FREQ{0:04}'.format(i)])
+    return np.array(c_freq_plane)
+
+
+def cosine_power_pattern(separation_rad, c_freq):
+    """
+    Power patterns for a given frequency based on the Cosine-squared power approximation
+    from Mauch et al. (2020).
+    Input: Radial separation array and a central frequency
+    """
+    rho = separation_rad
+    # Degrees to radians conversion
+    conv = np.pi / 180.
+    # converting arcminutes to radians
+    v_beam_rad = 89.5 / 60. * conv
+    h_beam_rad = 86.2 / 60. * conv
+    # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+    vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+    flux_density = []
+    for nu in c_freq:
+        nu = nu/1.e9
+        theta_b = vh_beam_mean*(1./nu)
+        ratio = rho/theta_b
+        num = np.cos(1.189*np.pi*ratio)
+        dem = 1-4*(1.189*ratio)**2
+        a_b = (num/dem)**2
+        flux_density.append(a_b)
+    return flux_density
+
+
+def beam_pattern(path):
+    """
+    Making beam pattern using the Cosine-squared power approximation from Mauch et al. (2020)
+    """
+    # Reading the fits file
+    data = read_fits(path)
+    # Getting pixel indices
+    row, col = np.indices((data.shape[2], data.shape[3]))
+    row = np.ravel(row)
+    col = np.ravel(col)
+    # Getting the central frequency of the image header given.
+    c_freq = central_freq(path)
+    # Getting radial separation between sources and the phase centre
+    phase_center, image_wcs = get_positions(path)
+    # Calling the radial_offset function
+    separation_rad = radial_offset(phase_center, image_wcs, row, col)
+    # Calling the cosine_power_pattern function
+    beam_list = cosine_power_pattern(separation_rad, c_freq)
+    return beam_list
+
+
+def mad(data):
+    """
+    Calculating median absolute deviation (MAD).
+    """
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    """
+    Calculating the standard deviation of each frequency plane using MAD. Rejecting pixels more
+    than 5 sigma from the mean until either no more pixels are rejected or a maximum  of 50
+    iterations is reached.
+    Returns: Weights per each frequency plane
+    """
+    diff = 1
+    i = 0
+    data = data[data != 0.0]
+    if len(data) == 0:
+        return float(0.0), float(0.0)
```

Comment:
Is the second return value intentional? All the other return statements in this program return a single value.